threaddb/buckets.go

```diff
@@ -66,6 +100,167 @@ type Deal struct {
 	Miner       string `json:"miner"`
 }
 
+// IsPrivate returns whether or not the bucket is private.
+func (b *Bucket) IsPrivate() bool {
+	return b.LinkKey != ""
+}
+
+// GetLinkEncryptionKey returns the bucket encryption key as bytes if present.
+// Version 0 buckets use the link key for all files and folders.
+// Version 1 buckets only use the link for folders.
+func (b *Bucket) GetLinkEncryptionKey() []byte {
+	return keyBytes(b.LinkKey)
+}
+
+// GetFileEncryptionKeyForPath returns the encryption key for path.
+// Version 0 buckets use the link key for all paths.
+// Version 1 buckets use a different key defined in path metadata.
+func (b *Bucket) GetFileEncryptionKeyForPath(pth string) ([]byte, error) {
+	if b.Version == 0 {
+		return b.GetLinkEncryptionKey(), nil
+	}
+
+	md, _, ok := b.GetMetadataForPath(pth, true)
+	if !ok {
+		return nil, fmt.Errorf("could not resolve path: %s", pth)
+	}
+	return keyBytes(md.Key), nil
+}
+
+func keyBytes(k string) []byte {
+	if k == "" {
+		return nil
+	}
+	b, _ := base64.StdEncoding.DecodeString(k)
+	return b
+}
+
+// GetFileEncryptionKeysForPrefix returns a map of keys for every path under prefix.
+func (b *Bucket) GetFileEncryptionKeysForPrefix(pre string) (map[string][]byte, error) {
```

Comment:
This is used to build a map of keys for _every_ metadata path under the given prefix. As noted above, any given key might be from a parent.